overlord/snapstate/snapstate.go

```diff
@@ -4146,6 +4147,121 @@ func GadgetConnections(st *state.State, deviceCtx DeviceContext) ([]gadget.Conne
+	if err != nil {
+		return nil, err
+	}
+
+	var revisionsToKeep []snap.Revision
+
+	// keep revisions in snap's sequence
+	for _, seq := range snapst.Sequence {
+		revisionsToKeep = append(revisionsToKeep, seq.Revision)
+	}
+
+	// keep revision in refresh hint (if it exists)
+	var refreshHints map[string]*refreshCandidate
+	if err := st.Get("refresh-candidates", &refreshHints); err != nil {
+		if !errors.Is(err, &state.NoStateError{}) {
+			return nil, fmt.Errorf("cannot get refresh-candidates: %v", err)
+		}
+	} else if hint, ok := refreshHints[snapName]; ok {
+		revisionsToKeep = append(revisionsToKeep, hint.Revision())
+	}
+
+	// keep revisions pointed to by a download task in an ongoing change
+	for _, chg := range st.Changes() {
+		if chg.IsReady() {
+			continue
+		}
+		for _, t := range chg.Tasks() {
+			if t.Kind() != "download-snap" {
+				continue
+			}
+			snapsup, _, err := snapSetupAndState(t)
+			if err != nil {
+				return nil, err
+			}
+			if snapsup.InstanceName() != snapName {
+				continue
+			}
+			revisionsToKeep = append(revisionsToKeep, snapsup.Revision())
+		}
+	}
+
+	return revisionsToKeep, nil
+}
+
+// cleanDownloads removes downloads that are no longer needed for all snaps.
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanDownloads = func(st *state.State) error {
+	snapStates, err := All(st)
+	if err != nil {
+		return err
+	}
+
+	keepContent := map[string]osutil.FileState{}
+	for snapName := range snapStates {
+		revisions, err := snapRevisionsToKeep(st, snapName)
+		if err != nil {
+			return err
+		}
+		for _, revision := range revisions {
+			mountFile := snap.MountFile(snapName, revision)
+			if osutil.FileExists(mountFile) {
+				keepContent[filepath.Base(mountFile)] = &osutil.FileReference{Path: mountFile}
+			}
+		}
+	}
+
+	_, _, err = osutil.EnsureDirState(dirs.SnapBlobDir, "*.snap", keepContent)
+	return err
+}
+
+// cleanSnapDownloads removes downloads that are no longer needed for the given snap.
+//
+// It is the caller's responsibility to lock state before calling this function.
+var cleanSnapDownloads = func(st *state.State, snapName string) error {
+	revisions, err := snapRevisionsToKeep(st, snapName)
+	if err != nil {
+		return err
+	}
+
+	keepContent := map[string]osutil.FileState{}
+	for _, revision := range revisions {
+		mountFile := snap.MountFile(snapName, revision)
+		if osutil.FileExists(mountFile) {
+			keepContent[filepath.Base(mountFile)] = &osutil.FileReference{Path: mountFile}
+		}
+	}
+
+	// XXX: could we need more digits?
+	const maxRevDigits = 10
+	var globs []string
+	for i := 1; i <= maxRevDigits; i++ {
+		var digitsPat string
+		for j := 1; j <= i; j++ {
+			digitsPat += "[0-9]"
+		}
+		globs = append(globs, fmt.Sprintf("%s_%s.snap", snapName, digitsPat))
+		globs = append(globs, fmt.Sprintf("%s_x%s.snap", snapName, digitsPat))
+	}
+
+	_, _, err = osutil.EnsureDirStateGlobs(dirs.SnapBlobDir, globs, keepContent)
```

Comment:
I think we can just use `fmt.Sprintf("%s_*.snap", snapName)`. AFAICT, that will pick up everything that we want it to. It's a bit broader but it seems fine given the pattern and the location. Wdyt?
```suggestion
    glob := fmt.Sprintf("%s_*.snap", snapName)
	_, _, err = osutil.EnsureDirState(dirs.SnapBlobDir, glob, keepContent)
```

Reply:
This is definitely much nicer but unfortunately parallel installs could be mistakenly deleted. for example this pattern `firefox_*.snap` would match the following for me.
```
firefox_2_3068.snap
firefox_2_3131.snap
firefox_3068.snap
firefox_3131.snap
```

Reply:
Ah because we use instance names and not snap names as I thought